docs/code-guidelines.md

```diff
@@ -0,0 +1,1141 @@
+    cwd, _ := os.Getwd()
+
+    // Bad: I/O
+    raw, _ := os.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+
+    yaml.Unmarshal(raw, &_config)
+}
+```
+
+- Good
+
+```go
+var _defaultFoo = Foo{
+    // ...
+}
+
+// or, better, for testability:
+
+var _defaultFoo = defaultFoo()
+
+func defaultFoo() Foo {
+    return Foo{
+        // ...
+    }
+}
+```
+
+```go
+type Config struct {
+    // ...
+}
+
+func loadConfig() Config {
+    cwd, err := os.Getwd()
+    // handle err
+
+    raw, err := os.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+    // handle err
+
+    var config Config
+    yaml.Unmarshal(raw, &config)
+
+    return config
+}
+```
+
+Considering the above, some situations in which `init()` may be preferable or necessary might include:
+
+- Complex expressions that cannot be represented as single assignments.
+- Pluggable hooks, such as `database/sql` dialects, encoding type registries, etc.
+- Optimizations to [Google Cloud Functions] and other forms of deterministic precomputation.
+
+  [Google Cloud Functions]: https://cloud.google.com/functions/docs/bestpractices/tips#use_global_variables_to_reuse_objects_in_future_invocations
+
+
+### Function Grouping and Ordering
+
+- Functions should be sorted in rough call order.
+- The receiver should group functions in a file.
+
+Therefore, exported functions should appear first in a file, after `struct`, `const`, and `var` definitions.
+
+A `newXYZ()`/`NewXYZ()` may appear after the type is defined but before the rest of the methods on the receiver.
+
+Since the receiver groups functions, plain utility functions should appear toward the end of the file.
+
+- Bad
+
+```go
+func (s *something) Cost() {
+  return calcCost(s.weights)
+}
+
+type something struct{ ... }
+
+func calcCost(n []int) int {...}
+
+func (s *something) Stop() {...}
+
+func newSomething() *something {
+    return &something{}
+}
+```
+
+- Good
+
+```go
+type something struct{ ... }
+
+func newSomething() *something {
+    return &something{}
+}
+
+func (s *something) Cost() {
+  return calcCost(s.weights)
+}
+
+func (s *something) Stop() {...}
+
+func calcCost(n []int) int {...}
+```
+
+
+### Reduce Nesting
+
+Code should reduce nesting where possible by handling error cases/special conditions first and returning early or continuing the loop. Reduce the amount of code that is nested on multiple levels.
+
+- Bad
+
+```go
+for _, v := range data {
+  if v.F1 == 1 {
+    v = process(v)
+    if err := v.Call(); err == nil {
+      v.Send()
+    } else {
+      return err
+    }
+  } else {
+    log.Printf("Invalid v: %v", v)
+  }
+}
+```
+
+- Good
+
+```go
+for _, v := range data {
+  if v.F1 != 1 {
+    log.Printf("Invalid v: %v", v)
+    continue
+  }
+
+  v = process(v)
+  if err := v.Call(); err != nil {
+    return err
+  }
+  v.Send()
+}
+```
+
+## Performance
+
+Performance-specific guidelines apply only to the hot path.
+
+#### Prefer strconv over fmt
+
```

Comment:
can we add 
#### avoid "+" for frequent string concatenation
- bad
```
func strConcat(n int, str string) string {
	s := ""
	for i := 0; i < n; i++ {
		s += str
	}

	return s
}
```

- good
a better option is  `strings.Builder` or `[]byte`
```
func strConcat(n int, str string) string {
	b := []byte{}
	for i := 0; i < n; i++ {
		buf = append(buf, str...)
	}

	return string(b)
}
```

In golang, string is immutable. when use "+", every time, there has to be a new memory allocation. For `[]byte`, the  memory of the slice []byte is allocated in multiples.

Reply:
Yes, this is a very nice one! sometimes the `+` can be more performative. But definitely is not so clear to read